clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/PubSubSplitPlanner.java

```diff
@@ -0,0 +1,111 @@
+ *       {@link com.linkedin.venice.vpj.pubsub.input.splitter.PartitionRangeSplitter}.</li>
+ * </ul>
+ */
+public class PubSubSplitPlanner {
+  private static final PubSubTopicRepository TOPIC_REPOSITORY = new PubSubTopicRepository();
+
+  public List<PubSubPartitionSplit> plan(VeniceProperties jobConfig) {
+    String brokerUrl = jobConfig.getString(KAFKA_INPUT_BROKER_URL);
+
+    try (TopicManager tm = createTopicManager(jobConfig, brokerUrl)) {
+      SplitType splitType =
+          SplitType.valueOf(jobConfig.getString(KAFKA_INPUT_SPLIT_TYPE, DEFAULT_KAFKA_INPUT_SPLIT_TYPE));
+      long recordsPerSplit =
+          jobConfig.getLong(KAFKA_INPUT_MAX_RECORDS_PER_MAPPER, DEFAULT_KAFKA_INPUT_MAX_RECORDS_PER_MAPPER);
+      int maxSplits = jobConfig.getInt(KAFKA_INPUT_MAX_SPLITS_PER_PARTITION, DEFAULT_MAX_SPLITS_PER_PARTITION);
+      long timeWindowMinutes =
+          jobConfig.getLong(KAFKA_INPUT_SPLIT_TIME_WINDOW_IN_MINUTES, DEFAULT_KAFKA_INPUT_TIME_WINDOW_IN_MINUTES);
+      long timeWindowMs = Duration.ofMinutes(timeWindowMinutes).toMillis();
+
+      PubSubTopic topic = TOPIC_REPOSITORY.getTopic(jobConfig.getString(KAFKA_INPUT_TOPIC));
+      return doPlan(tm, topic, splitType, recordsPerSplit, maxSplits, timeWindowMs);
+    }
+  }
+
+  @VisibleForTesting
+  TopicManager createTopicManager(VeniceProperties consumerProps, String brokerUrl) {
+    PubSubClientsFactory pubSubClientsFactory = new PubSubClientsFactory(consumerProps);
+    TopicManagerContext ctx = new TopicManagerContext.Builder().setPubSubPropertiesSupplier(k -> consumerProps)
+        .setPubSubTopicRepository(TOPIC_REPOSITORY)
+        .setPubSubPositionTypeRegistry(PubSubPositionTypeRegistry.fromPropertiesOrDefault(consumerProps))
+        .setPubSubAdminAdapterFactory(pubSubClientsFactory.getAdminAdapterFactory())
+        .setPubSubConsumerAdapterFactory(pubSubClientsFactory.getConsumerAdapterFactory())
+        .setTopicMetadataFetcherThreadPoolSize(1)
+        .setTopicMetadataFetcherConsumerPoolSize(1)
+        .setVeniceComponent(VeniceComponent.PUSH_JOB)
+        .build();
+    return new TopicManagerRepository(ctx, brokerUrl).getLocalTopicManager();
+  }
+
+  private List<PubSubPartitionSplit> doPlan(
+      TopicManager tm,
+      PubSubTopic topic,
+      SplitType splitType,
+      long recordsPerSplit,
+      int maxSplitsPerPartition,
+      long timeWindowMs) {
+    int partitionCount = RetryUtils.executeWithMaxAttempt(
+        () -> tm.getPartitionCount(topic),
+        10,
+        Duration.ofMinutes(1),
+        Collections.singletonList(Exception.class));
+
+    List<PubSubPartitionSplit> out = new ArrayList<>();
+    PartitionRangeSplitter splitter = PartitionRangeSplitter.getPartitionRangeSplitter(splitType);
+
+    for (int p = 0; p < partitionCount; p++) {
+      PubSubTopicPartition tp = new PubSubTopicPartitionImpl(topic, p);
+      SplitRequest req = new SplitRequest.Builder().pubSubTopicPartition(tp)
```

Comment:
trivial idea... we could have created a common builder with all fields except pubSubTopicPartition and used it to create new builder and add only tp to it